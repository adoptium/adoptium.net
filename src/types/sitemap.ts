export interface SitemapPage {
  title: string;
  url: string;
}

export interface SitemapSection {
  id: string;
  title: string;
  description?: string;
  pages: SitemapPage[];
}

export interface SitemapData {
  sections: SitemapSection[];
  dynamicContent?: {
    recentBlogs?: SitemapPage[];
  };
}

export const Sections = {
  topLevel: 'top-level',
  temurin: 'temurin',
  installation: 'installation',
  marketplace: 'marketplace',
  documentation: 'documentation',
  aqavit: 'aqavit',
  security: 'security',
  community: 'community',
  brand: 'brand',
  support: 'support',
} as const;

export type SectionId = typeof Sections[keyof typeof Sections];