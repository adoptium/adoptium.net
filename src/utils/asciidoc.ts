import Asciidoctor from "@asciidoctor/core";
import fs from "fs";
import path from "path";
import { decode } from "html-entities";

interface AsciidoctorDocument {
  getSourceLocation(): { getPath(): string };
}

interface AsciidoctorReader {
  pushInclude(
    content: string,
    file: string,
    target: string,
    line: number,
    attributes: Record<string, unknown>
  ): void;
}

interface AsciidoctorAttributes {
  [key: string]: unknown;
}

// Initialize Asciidoctor processor
const asciidoctor = Asciidoctor();

/**
 * Custom Asciidoctor include processor that resolves partials
 */
class CustomIncludeProcessor {
  private basePath: string;

  constructor(basePath: string) {
    this.basePath = basePath;
  }

  process(
    doc: AsciidoctorDocument,
    reader: AsciidoctorReader,
    target: string,
    attributes: AsciidoctorAttributes
  ) {
    // Get current file directory to resolve relative includes correctly
    const currentDir = path.dirname(doc.getSourceLocation().getPath());

    let includeFile = target;

    // If target is a relative path, resolve it from current file directory
    if (!path.isAbsolute(target) && !target.startsWith("/")) {
      includeFile = path.resolve(currentDir, target);
    }

    // If not found, try from the content base path
    if (!fs.existsSync(includeFile)) {
      includeFile = path.join(this.basePath, target);
    }

    if (fs.existsSync(includeFile)) {
      const content = fs.readFileSync(includeFile, "utf8");
      return reader.pushInclude(content, includeFile, target, 1, attributes);
    } else {
      console.warn(`Include file not found: ${includeFile}`);
      return reader.pushInclude(
        `[WARNING] Include file not found: ${target}`,
        target,
        target,
        1,
        attributes
      );
    }
  }
}

/**
 * Process AsciiDoc content with proper include resolution
 */
export function processAsciiDoc(
  filePath: string,
  content: string,
  options: Record<string, unknown> = {}
) {
  const basePath = path.dirname(filePath);
  const registry = asciidoctor.Extensions.create();

  // Create an instance of our custom processor
  const customProcessor = new CustomIncludeProcessor(basePath);

  // Try using a simple approach first - just skip extension registration
  // and see if the document still processes correctly
  const skipExtensions = true;

  if (!skipExtensions) {
    // Register include processor using a factory function
    const IncludeProcessor = function () {};
    IncludeProcessor.prototype.handles = function () {
      return true;
    };
    IncludeProcessor.prototype.process = function (
      doc: AsciidoctorDocument,
      reader: AsciidoctorReader,
      target: string,
      attrs: AsciidoctorAttributes
    ) {
      return customProcessor.process(doc, reader, target, attrs);
    };

    registry.includeProcessor(IncludeProcessor);
  }

  // Set default options with proper typing
  const defaultOptions: Record<string, unknown> = {
    safe: "server",
    base_dir: basePath,
    attributes: {
      "source-highlighter": "highlight.js",
      icons: "font",
      sectanchors: "",
      idprefix: "",
      idseparator: "-",
    },
  };

  // Add extension registry only if we're not skipping extensions
  if (!skipExtensions) {
    defaultOptions.extension_registry = registry;
  }

  // Merge options
  const mergedOptions = { ...defaultOptions, ...options };

  // Convert AsciiDoc to HTML - use the processor without extension registration
  // if we're having issues with extensions
  try {
    let html = asciidoctor.convert(content, mergedOptions) as string;
    // Post-process image src paths: replace src="filename.png" with src="/docs/filename.png" if not already absolute or external
    html = html.replace(
      /<img([^>]+)src=["'](?![a-z]+:|\/)([^"'>]+)["']/gi,
      '<img$1src="/docs/$2"'
    );
    return html;
  } catch (error) {
    console.error("Error converting AsciiDoc content:", error);

    // Try without extension registry as fallback
    // Create a new options object without the extension_registry property using destructuring
    // Use object destructuring with "rest" syntax to remove the property
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    const { extension_registry, ...fallbackOptions } = defaultOptions;
    let html = asciidoctor.convert(content, fallbackOptions) as string;
    html = html.replace(
      /<img([^>]+)src=["'](?![a-z]+:|\/)([^"'>]+)["']/gi,
      '<img$1src="/docs/$2"'
    );
    return html;
  }
}

/**
 * Decode HTML entities in a string
 */
export function decodeHtmlEntities(text: string): string {
  return decode(text);
}

/**
 * Extract metadata from AsciiDoc document
 */
export function extractMetadata(
  filePath: string,
  content: string
): {
  title: string;
  description: string;
  authors: string[];
  attributes: Record<string, unknown>;
} {
  const doc = asciidoctor.load(content, { safe: "server" });

  // Get the title and ensure it's a string
  const docTitle = doc.getDocumentTitle();
  let title =
    typeof docTitle === "string"
      ? docTitle
      : docTitle?.toString() || path.basename(filePath, path.extname(filePath));

  // Decode HTML entities in the title
  title = decodeHtmlEntities(title);

  // Get and decode description if it exists
  const description = doc.hasAttribute("description")
    ? decodeHtmlEntities(String(doc.getAttribute("description")))
    : "";

  return {
    title,
    description,
    authors: doc.hasAttribute("page-authors")
      ? String(doc.getAttribute("page-authors"))
          .split(",")
          .map((author: string) => author.trim())
      : [],
    attributes: doc.getAttributes() as Record<string, unknown>,
  };
}

/**
 * Check if a file exists
 */
export function fileExists(filePath: string): boolean {
  try {
    return fs.existsSync(filePath);
  } catch {
    // Silently handle any errors and return false
    return false;
  }
}

/**
 * Get all available language variants for a given base path
 */
export function getLanguageVariants(
  basePath: string,
  fileName: string
): string[] {
  const dir = path.dirname(basePath);
  const baseFileName = path.basename(fileName, ".adoc");

  // Get all adoc files in the directory
  const files = fs
    .readdirSync(dir)
    .filter((file) => file.startsWith(baseFileName) && file.endsWith(".adoc"));

  // Extract locales from filenames
  return files.map((file) => {
    // Format: index.de.adoc -> de
    const match = file.match(
      new RegExp(`${baseFileName}(?:\\.([\\w-]+))?\\.adoc$`)
    );
    return match && match[1] ? match[1] : "en";
  });
}
