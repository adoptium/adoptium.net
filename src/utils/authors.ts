import AuthorData from "@/data/authors.json";

export interface Author {
  name: string;
  summary?: string;
  twitter?: string;
  github?: string;
  linkedin?: string;
}

export function getFormattedAuthorData(authorId: string): Author {
  const rawAuthorData = AuthorData[authorId as keyof typeof AuthorData];

  if (!rawAuthorData) {
    return {
      name: "Unknown Author",
    };
  }

  // Convert the raw author data to our Author interface
  return {
    name: rawAuthorData.name || "Unknown Author",
    // Convert null values to undefined
    summary: rawAuthorData.summary || undefined,
    twitter: rawAuthorData.twitter || undefined,
    github: rawAuthorData.github || undefined,
    linkedin: rawAuthorData.linkedin || undefined,
  };
}
