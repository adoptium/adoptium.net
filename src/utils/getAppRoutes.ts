import fs from "fs";
import path from "path";
import matter from "gray-matter";

const LOCALE_DIR = path.join(process.cwd(), "src", "app", "[locale]");
const IGNORED_FOLDERS = ["__tests__"];

export function getAppRoutes(): string[] {
  const routes: string[] = [];

  function walk(dir: string, basePath = "") {
    const entries = fs.readdirSync(dir, { withFileTypes: true });

    for (const entry of entries) {
      const fullPath = path.join(dir, entry.name);
      const currentPath = path.join(basePath, entry.name);

      if (entry.isDirectory()) {
        if (IGNORED_FOLDERS.includes(entry.name)) continue;
        walk(fullPath, currentPath);
      }

      if (entry.name === "page.tsx") {
        let route =
          "/" +
          basePath
            .replace(/\/page\.tsx$/, "")
            .replace(/\/index$/, "")
            .replace(/\\/g, "/"); // Windows safety

        if (route === "/index" || route === "/") {
          route = "/";
        }

        routes.push(route);
      }
    }
  }

  walk(LOCALE_DIR);

  return routes;
}

interface BlogRoute {
  loc: string;
  lastmod: string;
}

export function getBlogRoutes(): BlogRoute[] {
  const blogDir = path.join(process.cwd(), "content", "blog");
  const folders = fs.readdirSync(blogDir, { withFileTypes: true });

  const routes: BlogRoute[] = [];

  for (const folder of folders) {
    if (!folder.isDirectory()) continue;

    const slug = folder.name;
    const indexPathMd = path.join(blogDir, slug, "index.md");
    const indexPathAdoc = path.join(blogDir, slug, "index.adoc");

    let indexPath = "";
    if (fs.existsSync(indexPathMd)) {
      indexPath = indexPathMd;
    } else if (fs.existsSync(indexPathAdoc)) {
      indexPath = indexPathAdoc;
    } else {
      continue;
    }

    // Read date from frontmatter
    let year = "";
    let month = "";
    try {
      const file = fs.readFileSync(indexPath, "utf8");
      const fm = matter(file);
      if (fm.data && fm.data.date) {
        const d = new Date(fm.data.date);
        year = d.getUTCFullYear().toString();
        month = (d.getUTCMonth() + 1).toString().padStart(2, "0");
      }
    } catch {}
    if (!year || !month) continue;

    const stats = fs.statSync(indexPath);
    const lastmod = stats.mtime.toISOString().split("T")[0];

    routes.push({
      loc: `/news/${year}/${month}/${slug}/`,
      lastmod,
    });
  }

  return routes;
}
