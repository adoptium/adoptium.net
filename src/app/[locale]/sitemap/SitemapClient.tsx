'use client';

import { useState, useMemo, useEffect } from 'react';
import { Link } from '@/i18n/navigation';
import { SitemapData, SitemapPage } from '@/types/sitemap';
import {
  FaHome,
  FaDownload,
  FaCog,
  FaStore,
  FaBook,
  FaCheckCircle,
  FaShieldAlt,
  FaUsers,
  FaPalette,
  FaHeart,
  FaNewspaper,
  FaChevronDown,
  FaChevronRight,
  FaSearch
} from 'react-icons/fa';
import { Sections } from '@/types/sitemap';

function useDebounce<T>(value: T, delay: number): T {
  const [debouncedValue, setDebouncedValue] = useState<T>(value);

  useEffect(() => {
    const handler = setTimeout(() => {
      setDebouncedValue(value);
    }, delay);

    return () => {
      clearTimeout(handler);
    };
  }, [value, delay]);

  return debouncedValue;
}

interface SectionWrapperProps {
  icon: React.ComponentType<{ className?: string }>;
  title: string;
  description?: string;
  children: React.ReactNode;
  isExpanded?: boolean;
  onToggle: () => void;
}

function SectionWrapper({ 
  icon: IconComponent, 
  title, 
  description, 
  children, 
  isExpanded = false, 
  onToggle 
}: SectionWrapperProps) {
  const headerContent = (
    <div className="flex items-center gap-4">
      <IconComponent className="text-2xl text-rose-500" />
      <div>
        <h2 className="text-xl font-semibold text-white">
          {title}
        </h2>
        {description && (
          <p className="text-sm text-gray-400 mt-1">
            {description}
          </p>
        )}
      </div>
    </div>
  );

  return (
    <div className="bg-gray-800/50 backdrop-blur-sm rounded-lg border border-gray-700 overflow-hidden">
        <button
          onClick={onToggle}
          className="w-full px-6 py-4 text-left hover:bg-gray-700/50 transition-colors duration-200 flex items-center justify-between"
        >
          {headerContent}
          {isExpanded ? (
            <FaChevronDown className="text-gray-400" />
          ) : (
            <FaChevronRight className="text-gray-400" />
          )}
        </button>

      {isExpanded && (
        <div className="px-6 pb-4 pt-2">
          {children}
        </div>
      )}
    </div>
  );
}

interface SitemapClientProps {
  sitemapData: SitemapData;
}

const sectionIcons: Record<string, React.ComponentType<{ className?: string }>> = {
  [Sections.topLevel]: FaHome,
  [Sections.temurin]: FaDownload,
  [Sections.installation]: FaCog,
  [Sections.marketplace]: FaStore,
  [Sections.documentation]: FaBook,
  [Sections.aqavit]: FaCheckCircle,
  [Sections.security]: FaShieldAlt,
  [Sections.community]: FaUsers,
  [Sections.brand]: FaPalette,
  [Sections.support]: FaHeart,
};

export default function SitemapClient({ sitemapData }: SitemapClientProps) {
  const [searchTerm, setSearchTerm] = useState('');
  const [expandedSections, setExpandedSections] = useState<Set<string>>(new Set());
  const [expandedDynamicSections, setExpandedDynamicSections] = useState<Set<string>>(new Set());

  const debouncedSearchTerm = useDebounce(searchTerm, 300);


  const filteredSections = useMemo(() => {
    if (!debouncedSearchTerm.trim()) return sitemapData.sections;

    const searchLower = debouncedSearchTerm.toLowerCase();
    return sitemapData.sections.map(section => ({
      ...section,
      pages: section.pages.filter(page =>
        page.title.toLowerCase().includes(searchLower)
      )
    })).filter(section => section.pages.length > 0);
  }, [sitemapData.sections, debouncedSearchTerm]);

  const filteredDynamicContent = useMemo(() => {
    if (!debouncedSearchTerm.trim() || !sitemapData.dynamicContent) {
      return sitemapData.dynamicContent;
    }

    const searchLower = debouncedSearchTerm.toLowerCase();
    return {
      recentBlogs: sitemapData.dynamicContent.recentBlogs?.filter(blog =>
        blog.title.toLowerCase().includes(searchLower)
      ),
    };
  }, [sitemapData.dynamicContent, debouncedSearchTerm]);

  const toggleSection = (sectionId: string) => {
    setExpandedSections(prev => {
      const newSet = new Set(prev);
      if (newSet.has(sectionId)) {
        newSet.delete(sectionId);
      } else {
        newSet.add(sectionId);
      }
      return newSet;
    });
  };

  const toggleDynamicSection = (sectionId: string) => {
    setExpandedDynamicSections(prev => {
      const newSet = new Set(prev);
      if (newSet.has(sectionId)) {
        newSet.delete(sectionId);
      } else {
        newSet.add(sectionId);
      }
      return newSet;
    });
  };

  const renderPageLink = (page: SitemapPage) => {
    const isExternal = /^https?:\/\//.test(page.url);
    const linkClasses = `block py-2 px-4 text-white hover:text-rose-600 transition-colors duration-200`;

    if (isExternal) {
      return (
        <a
          href={page.url}
          target="_blank"
          rel="noopener noreferrer"
          className={linkClasses}
        >
          {page.title}
        </a>
      );
    }

    return (
      <Link href={page.url} className={linkClasses}>
        {page.title}
      </Link>
    );
  };

  return (
    <div className="min-h-screen bg-purple">
      <div className="container mx-auto px-4 py-8">
        {/* Search Bar */}
        <div className="max-w-2xl mx-auto mb-8">
          <div className="relative">
            <FaSearch className="absolute left-4 top-1/2 transform -translate-y-1/2 text-gray-400" />
            <input
              type="text"
              placeholder="Search pages..."
              value={searchTerm}
              onChange={(e) => setSearchTerm(e.target.value)}
              className="w-full pl-12 pr-4 py-3 bg-gray-800 border border-gray-600 rounded-lg text-white placeholder-gray-400 focus:outline-none focus:ring-2 focus:ring-rose-500 focus:border-transparent"
            />
          </div>
        </div>

        {/* Sitemap Sections */}
        <div className="max-w-6xl mx-auto space-y-6">
          {filteredSections.map((section) => {
            const IconComponent = sectionIcons[section.id] || FaHome;
            const isExpanded = expandedSections.has(section.id);

            return (
              <SectionWrapper
                key={section.id}
                icon={IconComponent}
                title={section.title}
                description={section.description}
                isExpanded={isExpanded}
                onToggle={() => toggleSection(section.id)}
              >
                <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-2">
                  {section.pages.map((page) => (
                    <div key={page.url} className="group">
                      {renderPageLink(page)}
                    </div>
                  ))}
                </div>
              </SectionWrapper>
            );
          })}

          {/* Dynamic Content Sections */}
          {filteredDynamicContent && (
            <>
              {/* Recent Blogs */}
              {filteredDynamicContent.recentBlogs && filteredDynamicContent.recentBlogs.length > 0 && (
                <SectionWrapper
                  icon={FaNewspaper}
                  title="Latest News"
                  description="Recent blog posts and announcements"
                  isExpanded={expandedDynamicSections.has('recentBlogs')}
                  onToggle={() => toggleDynamicSection('recentBlogs')}
                >
                  <div className="grid grid-cols-1 md:grid-cols-2 gap-2">
                    {filteredDynamicContent.recentBlogs.map((blog) => (
                      <div key={blog.url} className="group">
                        {renderPageLink(blog)}
                      </div>
                    ))}
                  </div>
                </SectionWrapper>
              )}
            </>
          )}
        </div>
      </div>
    </div>
  );
}
