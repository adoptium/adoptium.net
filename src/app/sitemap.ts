import type { MetadataRoute } from "next";
import { getAppRoutes, getBlogRoutes } from "@/utils/getAppRoutes";
import { routing } from "@/i18n/routing";
import { getAllStaticSitemapUrls } from "@/utils/sitemapData";
import fs from "fs";
import path from "path";
import authorsData from "@/data/authors.json";

export default function sitemap(): MetadataRoute.Sitemap {
  const baseUrl = process.env.NEXT_PUBLIC_SITE_URL || "https://adoptium.net";

  const staticRoutes = getAppRoutes().filter(
    (route) => !route.includes("[...slug]") && !route.startsWith("/news")
  );

  // Getting additional routes from sitemap data to ensure all proposal URLs are included
  const sitemapUrls = getAllStaticSitemapUrls();
  const additionalRoutes = sitemapUrls
    .map(url => url.replace(/\/$/, '')) // Removing trailing slashes for consistency
    .filter(url => !staticRoutes.includes(url)); // Adding routes not already added in static routes

  const allStaticRoutes = [...staticRoutes, ...additionalRoutes];

  const blogRoutes = getBlogRoutes();

  // Collect asciidoc routes from content/asciidoc-pages (recursive)
  function getAsciidocRoutes(dir: string, prefix = ""): string[] {
    let routes: string[] = [];
    const entries = fs.readdirSync(dir, { withFileTypes: true });
    for (const entry of entries) {
      if (entry.isDirectory()) {
        routes = routes.concat(
          getAsciidocRoutes(
            path.join(dir, entry.name),
            `${prefix}/${entry.name}`
          )
        );
      } else if (entry.isFile() && entry.name.endsWith(".adoc")) {
        const route = `${prefix}/${entry.name.replace(/\index.adoc$/, "")}`;
        routes.push(route);
      }
    }
    return routes;
  }
  const asciidocDir = path.join(process.cwd(), "content/asciidoc-pages");
  let asciidocRoutes: string[] = [];
  try {
    asciidocRoutes = getAsciidocRoutes(asciidocDir);
  } catch {}

  const entries: MetadataRoute.Sitemap = [];

  // Build a set of canonical URLs handled by asciidoc index pages
  const asciidocCanonicalSet = new Set<string>();
  for (const route of asciidocRoutes) {
    const match = route.match(/(.+?)\/index(?:\.(\w{2,5}))?\.adoc$/);
    if (match) {
      const canonical = `${match[1]}/`;
      asciidocCanonicalSet.add(canonical);
    }
  }

  // Static pages (localized), but skip those handled by asciidoc index pages
  for (const route of allStaticRoutes) {
    const isRoot = route === "/";
    const canonical = isRoot ? "/" : route.endsWith("/") ? route : `${route}/`;
    if (!isRoot && asciidocCanonicalSet.has(canonical)) continue;
    const defaultUrl = isRoot ? `${baseUrl}/` : `${baseUrl}${route}`;
    const languages: Record<string, string> = {};
    for (const locale of routing.locales) {
      if (locale === routing.defaultLocale) continue;
      languages[locale] = isRoot
        ? `${baseUrl}/${locale}`
        : `${baseUrl}/${locale}${route}`;
    }
    // Add x-default hreflang pointing to the default locale
    if (Object.keys(languages).length > 0) {
      languages["x-default"] = defaultUrl;
    }
    entries.push({
      url: defaultUrl,
      lastModified: undefined,
      alternates: Object.keys(languages).length > 0 ? { languages } : undefined,
    });
  }

  // Blog/news pages (not localized, with date)
  for (const blog of blogRoutes) {
    const url = `${baseUrl}${blog.loc}`;
    let lastModified: Date | undefined = undefined;
    const match = blog.loc.match(/\/news\/(\d{4})\/(\d{2})/);
    if (match) {
      const [year, month] = match.slice(1);
      lastModified = new Date(Number(year), Number(month) - 1, 1);
    }
    entries.push({
      url,
      lastModified,
    });
  }

  // Author pages (localized)
  const authorRoutes = getAuthorRoutes();
  for (const route of authorRoutes) {
    const defaultUrl = `${baseUrl}${route}`;
    const languages: Record<string, string> = {};
    for (const locale of routing.locales) {
      if (locale === routing.defaultLocale) continue;
      languages[locale] = `${baseUrl}/${locale}${route}`;
    }
    // Add x-default hreflang pointing to the default locale
    if (Object.keys(languages).length > 0) {
      languages["x-default"] = defaultUrl;
    }
    entries.push({
      url: defaultUrl,
      lastModified: undefined,
      alternates: Object.keys(languages).length > 0 ? { languages } : undefined,
    });
  }

  // Asciidoc pages (localized by file extension)
  // Group by canonical path (parent dir) and collect language variants
  const asciidocMap: Record<string, Record<string, string>> = {};
  for (const route of asciidocRoutes) {
    // Match /foo/bar/index(.xx)?.adoc
    const indexMatch = route.match(/(.+?)\/index(?:\.(\w{2,5}))?\.adoc$/);
    if (indexMatch) {
      const canonical = `${indexMatch[1]}/`;
      const lang = indexMatch[2] || routing.defaultLocale;
      if (!asciidocMap[canonical]) asciidocMap[canonical] = {};
      asciidocMap[canonical][lang] = `${canonical}`;
      continue;
    }
    // Otherwise, treat as a regular page (no index)
    const match = route.match(/(.+?)(?:\.(\w{2,5}))?$/);
    if (!match) continue;
    const base = match[1];
    const lang = match[2] || routing.defaultLocale;
    if (!asciidocMap[base]) asciidocMap[base] = {};
    asciidocMap[base][lang] = `${base}${match[2] ? "." + match[2] : ""}`;
  }
  for (const base in asciidocMap) {
    const langs = asciidocMap[base];
    if (!langs[routing.defaultLocale]) continue;
    // Canonical URL: always ends with a slash
    let normalizedBase = base.startsWith("/") ? base : `/${base}`;
    normalizedBase = normalizedBase.replace(/\/+/g, "/");
    const defaultUrl = `${baseUrl}${normalizedBase}`;
    const languages: Record<string, string> = {};
    for (const locale of routing.locales) {
      if (locale === routing.defaultLocale) continue;
      if (langs[locale]) {
        languages[locale] = `${baseUrl}/${locale}${normalizedBase}`;
      }
    }
    // Add x-default hreflang pointing to the default locale
    if (Object.keys(languages).length > 0) {
      languages["x-default"] = defaultUrl;
    }
    entries.push({
      url: defaultUrl,
      lastModified: undefined,
      alternates: Object.keys(languages).length > 0 ? { languages } : undefined,
    });
  }

  // Collect author routes from authors.json
  function getAuthorRoutes(): string[] {
    try {
      return Object.keys(authorsData).map(
        (authorId) => `/news/author/${authorId}/`
      );
    } catch {
      return [];
    }
  }

  return entries;
}
