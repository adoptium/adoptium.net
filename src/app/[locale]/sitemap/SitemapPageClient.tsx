'use client';

import { useTranslations } from "next-intl";
import PageHeader from "@/components/Common/PageHeader";
import SitemapClient from "./SitemapClient";
import { SitemapData } from "@/types/sitemap";

interface SitemapPageClientProps {
  sitemapData: SitemapData;
}

export default function SitemapPageClient({ sitemapData }: SitemapPageClientProps) {
  const t = useTranslations("Sitemap");

  return (
    <div>
      <PageHeader
        title={t("title")}
        subtitle={t("subtitle")}
        description={t("description")}
      />
      <SitemapClient sitemapData={sitemapData} />
    </div>
  );
}
