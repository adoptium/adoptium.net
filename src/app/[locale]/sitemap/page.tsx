import { Metadata } from 'next';
import { generateSitemapData } from '@/utils/sitemapData';
import SitemapPageClient from './SitemapPageClient';

export const metadata: Metadata = {
  title: 'Sitemap',
  description: 'Complete sitemap of the Adoptium website including all pages, documentation, and resources.',
  robots: {
    index: true,
    follow: true,
  },
};

export default async function SitemapPage() {
  const sitemapData = generateSitemapData();

  return <SitemapPageClient sitemapData={sitemapData} />;
}
