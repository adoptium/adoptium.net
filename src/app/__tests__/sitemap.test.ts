import { describe, it, expect, beforeAll, vi, Mock } from "vitest";
import sitemap from "../sitemap";
import * as getAppRoutesModule from "@/utils/getAppRoutes";
import * as getBlogRoutesModule from "@/utils/getAppRoutes";

vi.mock("@/utils/getAppRoutes");
vi.mock("next/navigation", () => ({}));
vi.mock("next-intl/navigation", () => ({
  createNavigation: () => ({
    Link: () => null,
    redirect: () => {},
    usePathname: () => "",
    useRouter: () => ({}),
  }),
}));

const mockAppRoutes = ["/", "/page-1", "/page-2"];
const mockBlogRoutes = [
  { loc: "/news/2025/05/post-1", lastmod: "2025-05-09" },
  { loc: "/news/2024/12/other-post/", lastmod: "2024-12-01" },
];

describe("sitemap", () => {
  beforeAll(() => {
    (getAppRoutesModule.getAppRoutes as Mock).mockReturnValue(mockAppRoutes);
    (getBlogRoutesModule.getBlogRoutes as Mock).mockReturnValue(mockBlogRoutes);

    // Mock fs for asciidoc structure
    vi.mock("fs", () => {
      const files: Record<string, string[]> = {
        "/Users/foo/adoptium.net-next/content/asciidoc-pages": [
          "about",
          "docs",
        ],
        "/Users/foo/adoptium.net-next/content/asciidoc-pages/about": [
          "index.adoc",
          "index.de.adoc",
        ],
        "/Users/foo/adoptium.net-next/content/asciidoc-pages/docs": [
          "installation.adoc",
        ],
      };
      return {
        readdirSync: (dir: string) => files[dir] || [],
        existsSync: (path: string) => {
          // All files and dirs above exist
          return (
            Object.keys(files).includes(path) ||
            Object.values(files).some((arr) =>
              arr.includes(path.split("/").pop()!)
            )
          );
        },
        statSync: (path: string) => ({
          isDirectory: () => Object.keys(files).includes(path),
        }),
        readFileSync: (path: string) => {
          // Return fake frontmatter for blog/asciidoc if needed
          if (path.endsWith("installation.adoc")) {
            return "---\ntitle: Installation\n---\nContent";
          }
          if (path.endsWith("index.adoc")) {
            return "---\ntitle: About\n---\nContent";
          }
          if (path.endsWith("index.de.adoc")) {
            return "---\ntitle: Ãœber\n---\nContent";
          }
          return "";
        },
      };
    });
  });

  it("generates the expected sitemap snapshot", () => {
    const result = sitemap();
    expect(result).toMatchSnapshot();
  });
});
