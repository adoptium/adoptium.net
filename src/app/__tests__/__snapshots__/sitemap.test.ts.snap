// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`sitemap > generates the expected sitemap snapshot 1`] = `
[
  {
    "alternates": {
      "languages": {
        "de": "https://adoptium.net/de",
        "en-GB": "https://adoptium.net/en-GB",
        "es": "https://adoptium.net/es",
        "fr": "https://adoptium.net/fr",
        "pt-BR": "https://adoptium.net/pt-BR",
        "zh-CN": "https://adoptium.net/zh-CN",
      },
    },
    "lastModified": undefined,
    "url": "https://adoptium.net/",
  },
  {
    "alternates": {
      "languages": {
        "de": "https://adoptium.net/de/page-1",
        "en-GB": "https://adoptium.net/en-GB/page-1",
        "es": "https://adoptium.net/es/page-1",
        "fr": "https://adoptium.net/fr/page-1",
        "pt-BR": "https://adoptium.net/pt-BR/page-1",
        "zh-CN": "https://adoptium.net/zh-CN/page-1",
      },
    },
    "lastModified": undefined,
    "url": "https://adoptium.net/page-1",
  },
  {
    "alternates": {
      "languages": {
        "de": "https://adoptium.net/de/page-2",
        "en-GB": "https://adoptium.net/en-GB/page-2",
        "es": "https://adoptium.net/es/page-2",
        "fr": "https://adoptium.net/fr/page-2",
        "pt-BR": "https://adoptium.net/pt-BR/page-2",
        "zh-CN": "https://adoptium.net/zh-CN/page-2",
      },
    },
    "lastModified": undefined,
    "url": "https://adoptium.net/page-2",
  },
  {
    "lastModified": 2025-05-01T00:00:00.000Z,
    "url": "https://adoptium.net/news/2025/05/post-1",
  },
  {
    "lastModified": 2024-12-01T00:00:00.000Z,
    "url": "https://adoptium.net/news/2024/12/other-post/",
  },
]
`;
